### PR TITLE
HighKingSafety

### PR DESCRIPTION
**Pull request by NicklasPersson**
STC

ELO: 2.88 +-2.2 (95%) LOS: 99.5%
Total: 40000 W: 8312 L: 7980 D: 23708

LTC

LLR: 2.96 (-2.94,2.94) [0.00,4.00]
Total: 11529 W: 2075 L: 1882 D: 7572

bench 8449970

### CODE REVIEW

**Comment by glinscott** [comment]
Very impressive!!  Change looks good.

File: src/evaluate.cpp [hunk 1]
```diff
@@ -411,11 +411,11 @@ namespace {
...
-        attackUnits =  std::min(77, ei.kingAttackersCount[Them] * ei.kingAttackersWeight[Them])
-                     + 10 * ei.kingAdjacentZoneAttacksCount[Them]
-                     + 19 * popcount<Max15>(undefended)
-                     +  9 * (ei.pinnedPieces[Us] != 0)
-                     - mg_value(score) * 63 / 512
+        attackUnits =  std::min(74, ei.kingAttackersCount[Them] * ei.kingAttackersWeight[Them])
+                     + 8 * ei.kingAdjacentZoneAttacksCount[Them]
+                     + 25 * popcount<Max15>(undefended)
+                     +  10 * (ei.pinnedPieces[Us] != 0)
+                     - mg_value(score) * 62 / 512
```
Comment (hxim) [inline src/evaluate.cpp]: This is same as  "- mg_value(score) * 31 / 256"

Comment (NicklasPersson) [inline src/evaluate.cpp]: Thanks

**Comment by lucasart** [comment]
Good stuff. Only thing is that it's hard for me to intuitively understand why knight attack weight is worth 4x more than bishop or 8x more than queen. But, the proof is in the pudding, and this tuning is a clear elo gain.

**Comment by DU-jdto** [comment]
My guess is that it's something to do with the fact that when a bishop or queen isn't involved in the attack, it can often maneuver to be so much more quickly than an uninvolved knight could. That plus the fact that there is already a significant king safety penalty for the presence of an enemy queen on the board.

**Comment by joergoster** [comment]
@NicklasPersson Congrats! Great work!
@lucasart Very often a double attack or even a fork is involved, when the knight is attacking. The enemy is urged to take, and so opens up a file or diagonal for the sliding pieces. Furthermore, an attack by a sliding piece might be interrupted by a simple pawn move or a piece. This is not possible with a knight attack!

**Comment by Karlon** [comment]
Looks good.

**Comment by NicklasPersson** [comment]
@lucasart,@joergoster,@DU-jdto: or perhaps the ELO-gain is somewhere else and the KingAttackWeights changes might for all I know be a ELO-drop. Remember the linear extrapolation I used, created from the SPSA-sessions, is very crude and parameters with small absolute value will have a larger relative change. Due to this and how ridiculously far down the line I could go (much farther than I could imagine)  there are probably several parameters which are badly tuned and needs an SPSA-session which I plan to schedule this evenning.

**Comment by Karlon** [comment]
Nicklass, I have spent half an hour fuzzing around with this patch and have reached a solid conclusion; It's both an elo gain, and an elo loss. The elo gain is 2+ like you mentioned above, but the elo loss is approximately 2 elo too! So I suggest that this patch gets safely neglected and instead, wait for Nicklass to come up with another one that does not have any sort of adverse essence.

**Comment by Karlon** [comment]
I do agree that the scenary I showed above, is quite exuberant and vivid in colour. It exhibits the cardial gist of natural phenomena.

**Comment by lucasart** [comment]
@Karlon: stop spamming. Thanks!

**Comment by lantonov** [comment]
I thought that the "adverse essence" was banned by Marco ?!

**Comment by ajithcj** [comment]
Probably it is better to re-run below test at LTC SPRT[0,4] :
http://tests.stockfishchess.org/tests/view/54bebaa00ebc5946cd385822

If it passes a separate pull request can be raised with those values instead of this one. Better to use SPSA tuned values instead of manually-tuned values.

**Comment by ajithcj** [comment]
I mean the LTC test should be run between
KingSafetySPSA (King safety tuning with values obtained by SPSA.) vs Master(SF6 RC 1)

**Comment by zamar** [comment]
Looks good. Stuff for SF7.

**Comment by mcostalba** [comment]
@NicklasPersson you can close this, it seems superseded by #228

**Comment by NicklasPersson** [comment]
Okay.

**Pull request closed without merging**